---
permalink: /nabishop/
title: "Nicholas Bishop"
excerpt: "CSC 448 Bioinformatics Algorithms Project"
---

## Introduction
The following page is an attempt to help the cause of understanding and curing the COVID-19 virus that is currently sweeping the world. Each project section of the focus of a given week, with some being more focused on synthesizing data and research and others being focused on using algorithms to make more in-depth conclusions. Both approaches attempt at helping the medical field for fighting COVID-19.

## Project 1
For the first week of the project, I will review several available resources which discuss and analyze the COVID-19 virus. Each subsequent section will break down a given source, with of course, credit to where the information came from.

### GISAID
The GISAID Initiative was founded in 2006 and helps share global information of influenza virus sequences to help professionals better understand the growth and evolution of virus strands. After signing up, anyone can access the public data. On the organization's website, there are several interactive data plots to a variety of COVID-19 samples including those with amino acid mutations and country of sample as well as a growth map of how the COVID-19 virus spread across the world and which areas are most impacted. 

GISAID information can be found at https://www.gisaid.org/

### Folding@home
Folding@home is a computing project that is made for simulating the process of movement and folding of proteins that are part of forms of illnesses. To help with simulations, anyone can volunteer processing power from their computers for scientists and researchers to use for more processing power in simulating protein interactions and behavior, including that of COVID-19. Outside of viruses, the organization does a significant amount of research on other diseases such as cancer, Alzheimer's disease, and Chagas disease.

Folding@home information can be found at https://foldingathome.org/

### Nextstrain
Nextstrain is an open-source project that provides tools for visualizing and analyzing pathogen genome information using publicly available data. In their epidemiology of COVID-19, Nextstrain utilizes data from GISAID, mentioned above, to visualize the growth and evolution of the virus since its first documented case in Wuhan, China. There are many different ways to visualize the genome evolution and documentation by having tree options, date ranges, and several other filters like narrowing results down to a given country to help researchers more easily find relevant information.

Nextstrain information can be found at https://nextstrain.org/ncov

### Complete Genome Sequence of a 2019 Novel Coronavirus (SARS-CoV-2) Strain Isolated in Nepal article by American Society For Microbiology
The article by the American Society For Microbiology describes the rising spread of the COVID-19 virus and breaks down its genome. The given genome was sequenced from a 32-year old Nepalese man who studied at the Wuhan University of Technology in Wuhan, China. The full RNA genome was directly extracted using gene-specific primers and the Burrows-Wheeler Aligner MEM algorithm (BWA-MEM) 0.7.5a-r405 assembly method. The findings showed that the COVID-19 virus is a single, positive-stranded RNA that is 29,811 nucleotides long and is composed of 29.86% adenosines, 18.39% cytosines, 19.63% guanines, and 32.12% thymines.

The article can be found at https://mra.asm.org/content/9/11/e00169-20

### COVID-19 Analysis Using Galaxy and BioConda
Using Galaxy, an open platform for genome computation research, and BioConda, an Anaconda package manager that specializes in bioinformatics technology, analysis was done on COVID-19, specifically regarding the virus’s genome and evolution. Current genome research includes documenting the virus’s makeup and intrahost variation using 33 publicly available samples. Data used in analyzing the evolution of the virus has narrowed down five potential genomic positions that may be a location for positive or negative selection, which is incredibly important for understanding the adaptations of COVID-19. 

The COVID-19 analysis can be found at https://covid19.galaxyproject.org/
